### ISSUE TITLE
Multiple roots and items that are not files

### ISSUE THREAD

**Issue by LoganDark** [NeekSandhu/aspen-core#14]
Multiple roots and special items (like "buttons" at the bottom, or section separators, or an item at the top of the list describing the root directory, etc.) can be added by essentially making a virtual filesystem using the `IBasicFileSystemHost`. However, this is a lot of effort and it would be nice to know other solutions to the problem.

Does anything like this currently exist? If not, do you have plans to add any?

**Comment by NeekSandhu** [comment]
Very reasonable and a valid feature request.

The next major release of `react-aspen` (in about a month or 2) will have 100% API compliance with [VSCode's TreeView API](https://code.visualstudio.com/api/extension-guides/tree-view). It's as raw as it gets, meaning the `TreeDataProvider` is agnostic of what "type" of data it is, thus you will be able to make a tree of anything.

Although, I must agree that a multi-root tree should be a built-in feature, but no timeline on that.

For now, you can achieve this by "shimming" your host like so:

```ts

const host: IBasicFileSystemHost = {
	pathStyle: 'unix', // or 'win32'
	getItems: (path: string): IFileEntryItem[] => {
		if (path === '/') {
          return [
            {
              name: '/home/logandark/folder1',
              type: FileType.Directory
            },
            {
              name: '/etc/whatever',
              type: FileType.Directory
            }
          ]
        }
        
        // return fs.readdirSync(path).map(...);
	}
}

const treeModel: TreeModel = new TreeModel(host, '/')

```

With the code above, the first call to `getItems` when the `TreeModel` is booting up will set your "multi-root workspace". Adjust your item renderer to render the label as you deem fit.

Note that you can also leverage the `IFileEntryItem#metadata` field to your advantage and use it to add custom information to it. This metadata will be made available to you on the other side when that item will be rendered. For example, if you want to achieve separators.

___

The buttons that you propose probably don't belong in the tree structure or even aspen for that matter, they are more like a footer content:

```tsx
<Sidebar>
  <FileTree ...>
  <Footer>
    <NewFileButton />
    <NewFolderButton />
  </Footer>
</Sidebar>
```

**Comment by LoganDark** [comment]
> The next major release of `react-aspen` (in about a month or 2) will have 100% API compliance with [VSCode's TreeView API](https://code.visualstudio.com/api/extension-guides/tree-view).

I'm kinda sad that you're using VSCode stuff verbatim, especially since their Provider/Thenable stuff seems way too overengineered - I appreciate the simple Promise-based API a **lot** more, and likely would not want to use your library if it seemed like it was trying way too hard to make it into Monaco/VSC. I would highly encourage you to re-think this description and choose a solution that would better fit this library and ease of use rather than integration into Microsoft products.

> For now, you can achieve this by "shimming" your host like so:
> 
> ```ts
> const host: IBasicFileSystemHost = {
> 	pathStyle: 'unix', // or 'win32'
> 	getItems: (path: string): IFileEntryItem[] => {
> 		if (path === '/') {
>           return [
>             {
>               name: '/home/logandark/folder1',
>               type: FileType.Directory
>             },
>             {
>               name: '/etc/whatever',
>               type: FileType.Directory
>             }
>           ]
>         }
>         
>         // return fs.readdirSync(path).map(...);
> 	}
> }
> 
> const treeModel: TreeModel = new TreeModel(host, '/')
> ```
> 
> With the code above, the first call to `getItems` when the `TreeModel` is booting up will set your "multi-root workspace". Adjust your item renderer to render the label as you deem fit.

Will this work on Windows with drive letters? If I provide, say, `\\` as the root for the TreeModel, and then return `C:\\...` and `D:\\...` as extra "roots", would that work?
> they are more like a footer content:

I did consider this. I'll have to look into it to make sure the virtual list stuff won't conflict with it.

**Comment by NeekSandhu** [comment]
> I'm kinda sad that you're using VSCode stuff verbatim, especially since their Provider/Thenable stuff seems way too overengineered - I appreciate the current simple Promise-based API a lot more, and likely will not want to use the next release if it turns into something trying way too hard to make it into Monaco/VSC

Don't you worry. The current API's will always stay as-is. It'll be implemented such that `FileTree` component's `model` prop will accept either `instanceof TreeModel` or `instanceof TreeDataProvider`, the former being what's currently supported. And I'll be skipping their "proprietary" Thenable/Promise mechanisms.

> Will this work on Windows with drive letters? If I provide, say, \\ as the root for the TreeModel, and then return C:\\... and D:\\... as extra "roots", would that work?

I'm not even certain if this will work with unix paths. You can give it a shot though, I'd love to hear your findings.

___

> Although, I must agree that a multi-root tree should be a built-in feature, but no timeline on that.

Once you spend some time tinkering with aspen ecosystem, perhaps you can share some ideas around how this can be done elegantly based on your experience with whatever you're building.

**Comment by LoganDark** [comment]
> It'll be implemented such that `FileTree` component's `model` prop will accept either `instanceof TreeModel` or `instanceof TreeDataProvider`, the former being what's currently supported.

What if the Microsoft APIs were supported by an optional add-on (`aspen-vsc` perhaps?)? That way all of its features will be supported by Aspen, since the add-on would have to use the base APIs to implement the Microsoft ones. That would also mean people like me who are paranoid and afraid of dumb Microsoft crap won't have to pull all of that baggage/dead code in with the main library.

> I'm not even certain if this will work with unix paths. You can give it a shot though, I'd love to hear your findings.

I'm sure Unix paths will be fine, especially if path-fx's join mechanism (or whatever you're using to get full paths) handles double slashes correctly. It's just a question of whether aspen handles slashes in ""filenames"" correctly. I know that Windows does not recognize `\\C\\`, so this will be an interesting experiment.

> Once you spend some time tinkering with aspen ecosystem, perhaps you can share some ideas around how this can be done elegantly based on your experience with whatever you're building.

I'll get back to you on that if I have any ideas.

**Comment by NeekSandhu** [comment]
> I've just tested this and it appears that your library is too smart. Aspen strips everything but the basename and then errors trying to read some random directory in the root of `C:`.

Ah yes! I just remembered why it does that. This to maintain intrinsic integrity of "information". It would not be "legal" to have `FileEntry` of path `"/home/defunkt/desktop/projects/timer"` somehow being a child of `"/home/defunkt/downloads"` in the tree view.

I'll give some thought on making multi-root tree a native feature in the next release. But for now, my personal suggestion would be to have your multi-roots in sandboxed `FileTree` instances, and then use accordions to emulate different roots:

```jsx
    <Sidebar>
        <Accordion>
            <AccordionItem>
                home/logandark/folder1
                <FileTree model={modelBoundToFolder1} />
            </AccordionItem>
            <AccordionItem>
				etc/whatever
                <FileTree model={modelBoundToWhatever} />
            </AccordionItem>
        </Accordion>
    </Sidebar>
```

See `react-accessible-accordion`

**Comment by LoganDark** [comment]
I plan on implementing shift+click to select as well as arrow keys and other navigation features, and using multiple FileTrees like that will complicate implementation. I'll probably just go with filesystem virtualization (let aspen think that `\path1\Documents` exists, while in reality it reads from e.g. `C:\Users\LoganDark\Documents`)

**Comment by NeekSandhu** [comment]
> I'll probably just go with filesystem virtualization (let aspen think that \path1\Documents exists, while in reality it reads from e.g. C:\Users\LoganDark\Documents)

That's actually a pretty decent workaround for the time being.

I'll still leave the issue open for reference and will be closed along with next major release

**Comment by LoganDark** [comment]
> What if the Microsoft APIs were supported by an optional add-on (`aspen-vsc` perhaps?)? That way all of its features will have to be supported by Aspen, since the add-on would have to use the base APIs to implement the Microsoft ones, but the base API can stay sane. That would also mean people like me who are paranoid and afraid of dumb Microsoft crap won't have to pull all of that baggage/dead code in with the main library.
> 
> Perhaps going with a monorepo structure, as you said that you want to do, would help with this. Still same repo, but different packages.

Have you made up your mind about this yet? I guess by now it would be too late if you've already started implementing the special Microsoft APIs as-is in the core...

**Comment by zikaari** [comment]
VSCode API adaptor will not be part of the core _(even if it is, it'll be tree-shakable)_. To clarify, this is how it'll work - `aspen-core` is the real backend of `react-aspen`. The former keeps the low-level parts in sync once they are spawned until they are destroyed. The low-level parts are `Directory` and `FileEntry`. How those low-level parts are born is decided by the adaptors. Currently, you have `IBasicFileSystemHost` which maps the returned array from `getItems` to those low-level things and are then handed off to `aspen-core`.

Aspen 2 will just ship with another adaptor that just like `IBasicFileSystemHost`, will also support the VSCode interface, that is `TreeDataProvider`. The reason for adding this support is to make Aspen more general purpose, for example when you don't have a filesystem and paths to work with. The end-user can either use `IBasicFileSystemHost`, i.e the classic API, OR, `TreeDataProvider`, whichever works best for them.

Let me know your concerns that you see with this change and I will be happy to answer.

**Comment by LoganDark** [comment]
> Let me know your concerns that you see with this change and I will be happy to answer.

That sounds pretty close to what I specified above, but my concern is that you seem to be saying that the vscode interface is the only way to represent a non-filesystem tree easily? Like somehow that has to abstract over some hacky filesystem stuff underneath that the core aspen API suffers from? Shouldn't you be solving that in the aspen 2 release?

**Comment by zikaari** [comment]
> Shouldn't you be solving that in the aspen 2 release

Yes, Aspen 2 will be the one with this change.

> abstract over some hacky filesystem stuff underneath that the core aspen API suffers from

This is exactly what'll be decoupled. Currently, when I say `IBasicFileSystemHost` is an adaptor, it's not 100% true because it's actually part of `aspen-core`. They are deeply connected to the point for `aspen-core` literally relies on paths and whatnot. This is what'll be refactored out such that `aspen-core` itself is agnostic of paths, renames, deletions, and additions. Instead `aspen-core` will have the adaptors implement the necessary interface to carry out the operations, i.e renames, deletions and additions. Whether those operations are dependent on paths, or something else, `aspen-core` doesn't care. So the "knowledge" of paths will just be moved from `aspen-core`'s core over to the respective adaptors, in this case, `IBasicFileSystemHost`. It'll be an internal refactoring that end-users need not worry about.

**Comment by LoganDark** [comment]
> > abstract over some hacky filesystem stuff underneath that the core aspen API suffers from
> 
> This is exactly what'll be decoupled. Currently, when I say `IBasicFileSystemHost` is an adaptor, it's not 100% true because it's actually part of `aspen-core`. They are deeply connected to the point for `aspen-core` literally relies on paths and whatnot. This is what'll be refactored out such that `aspen-core` itself is agnostic of paths, renames, deletions, and additions. Instead `aspen-core` will have the adaptors implement the necessary interface to carry out the operations, i.e renames, deletions and additions. Whether those operations are dependent on paths, or something else, `aspen-core` doesn't care. So the "knowledge" of paths will just be moved from `aspen-core`'s core over to the respective adaptors, in this case, `IBasicFileSystemHost`. It'll be an internal refactoring that end-users need not worry about.

Will it become easier to represent multiple roots and non-file items and such without having to use the VSCode package? What will happen to `IBasicFileSystemHost` - will it become its own package? Or do you mean that the responsibility of dealing with paths will move to implementers of `IBasicFileSystemHost`?

**Comment by zikaari** [comment]
> Will it become easier to represent multiple roots and non-file items and such without having to use the VSCode package

Yes. I plan on doing this by **not** asking for root path while spawning a tree model.

```diff
- new TreeModel(host, '/absolute/path/that/will/act/as/root')
+ new TreeModel(host)
```

Instead, the root(s) will be decided during the first-ever call to `getItems` when it'll be called with a `null` path argument. At that time, you are welcome to either return the contents of `/etc`, or multiple root(s) instead, i.e `['/etc', '/user']`. What you return here will simply become the first level items that you see in the tree.

> non-file items and such

I'm not sure how this will actually work. Do you have any ideas?

> will it become its own package

No. Very likely a tree shakable export.

> Or do you mean that the responsibility of dealing with paths will move to implementers of `IBasicFileSystemHost`?

No, that won't be happening, `inotify` will continue to live as it always has.

**Comment by LoganDark** [comment]
> Instead, the root(s) will be decided during the first-ever call to `getItems` when it'll be called with a `null` path argument. At that time, you are welcome to either return the contents of `/etc`, or multiple root(s) instead, i.e `['/etc', '/user']`. What you return here will simply become the first level items that you see in the tree.
> 
> >

And what if I want to add a new one? What would I pass to inotify? A top-level directory entry with a name that starts with a slash?

> I'm not sure how this will actually work. Do you have any ideas?

I mean, my best idea would probably be to make it so that paths don't necessarily have to be strings, and aspen is generic over the type of path. But that would probably require a rewrite of the entire library.

**Comment by zikaari** [comment]
> And what if I want to add a new one? What would I pass to inotify?

Just call `inotify` with `IWatcherAddEvent` and set `directory` property to `null`. This will imply the change is at the top-most level.

> I mean, my best idea would probably be to make it so that paths don't necessarily have to be strings, and aspen is generic over the type of path

Perhaps during the re-write, I'll find some light.

**Comment by zikaari** [comment]
Can you give me bunch of example use-cases of non-file items within the tree?

**Comment by LoganDark** [comment]
> Can you give me bunch of example use-cases of non-file items within the tree?

Debugger - listing local variables and letting them be explored

Output console - grouping identical log messages

Exploring JSON objects [or other kinds of objects]

Most of these would be difficult to emulate with string file paths (well, maybe not the JSON object one)

**Comment by zikaari** [comment]
Yes. All of the cases are the top priority of Aspen 2. I thought you wanted to have non-file items within a "file tree".

**Comment by LoganDark** [comment]
> I thought you wanted to have non-file items within a "file tree".

That's what I have to do right now, yes. There is a special folder that contains metadata files for special panels that the user has open, and I turn those metadata files into entries that open the relevant panels when clicked.

**Comment by zikaari** [comment]
Taken from Aspen 2's draft changelog -

> `BasicFileSystemHost` has been removed in favor of a generic and more performant `TreeSource` provider. Where you'd previously implement a `getItems` method that would return an array of file entry descriptors, there is now `getNodes` instead. The `getNodes` method will be called with a literal `Branch` and you'll return an array of `TreeNode`s for it. No more "descriptor" objects, instead the actual `Branch`s and `Leaf`s using very simple factory functions to create either of those.

With this, you'd have code that looks like this:

```tsx
	type FileEntry = {
		path: string;
	};
 
	const tree = new Tree<FileEntry>({
		getNodes: (parent: Branch<FileEntry>, { createBranch, createLeaf }: TreeNodeFactory<FileEntry>) => {
			if (!parent) {
				return [
					createBranch({ path: 'C:\\Users\\logandark\\Downloads' }),
					createBranch({ path: 'D:\\Steam\\apps' })
				];
			}
			
			return (await fs.readdir(parent.data.path)).map(fileEntry => fileEntry.isDirectory()
				? createBranch({ path:  Path.join(parent.data.path, fileEntry.name) })
				: createLeaf({ path:  Path.join(parent.data.path, fileEntry.name) })
			);
		}
	});
```

Let me know your thoughts?

Also, can you share what kind of product/app are you building using Aspen?

**Comment by LoganDark** [comment]
> Taken from Aspen 2's draft changelog -
> 
> > `BasicFileSystemHost` has been removed in favor of a generic and more performant `TreeSource` provider. Where you'd previously implement a `getItems` method that would return an array of file entry descriptors, there is now `getNodes` instead. The `getNodes` method will be called with a literal `Branch` and you'll return an array of `TreeNode`s for it. No more "descriptor" objects, instead the actual `Branch`s and `Leaf`s using very simple factory functions to create either of those.
> 
> With this, you'd have code that looks like this:
> 
> ```tsx
> 	type FileEntry = {
> 		path: string;
> 	};
>  
> 	const tree = new Tree<FileEntry>({
> 		getNodes: (parent: Branch<FileEntry>, { createBranch, createLeaf }: TreeNodeFactory<FileEntry>) => {
> 			if (!parent) {
> 				return [
> 					createBranch({ path: 'C:\\Users\\logandark\\Downloads' }),
> 					createBranch({ path: 'D:\\Steam\\apps' })
> 				];
> 			}
> 			
> 			return (await fs.readdir(parent.data.path)).map(fileEntry => fileEntry.isDirectory()
> 				? createBranch({ path:  Path.join(parent.data.path, fileEntry.name) })
> 				: createLeaf({ path:  Path.join(parent.data.path, fileEntry.name) })
> 			);
> 		}
> 	});
> ```
> 
> Let me know your thoughts?

Oh I love the new approach, it very easily generalizes to more than just filesystems. I haven't reviewed the full draft yet.

> Also, can you share what kind of product/app are you building using Aspen?

I can't really reveal that at the moment. However I can say for certain that it is a filesystem view - just not a straight view of a single folder's contents.
